docs/design/proposals/rbd-mirror.md

```diff
@@ -0,0 +1,73 @@
+# RBD MIRRORING
+
+RBD mirroring is a process of replication of RBD images between two or more
+Ceph clusters. Mirroring ensures point-in-time, crash-consistent RBD images
+between clusters, RBD mirroring is mainly used for disaster recovery (i.e.
+having a secondary site as a failover). See Upstream [Ceph
+documentation](https://docs.ceph.com/en/latest/rbd/rbd-mirroring) on RBD
+mirroring for complete information.
+
+## Architecture
+
+![mirror](rbd-mirror.png)
+
+## Design
+
+Currently, ceph-csi generates its unique ID for each RBD image and stores the
+corresponding PV name and the unique ID mapping, It creates the RBD image with
+the unique ID and returns the encoded value which contains all the required
+information for other operations. For mirroring, the same RBD image will be
+mirrored to the secondary cluster. As the journal(omap data) is not mirrored to
+the secondary cluster, The cephcsi cannot identify the RBD image corresponding
+to the PV wihtout omap data.
+
+**Note:** It's expected that the user/admin will create the static PVC's on the
+secondary site during the failover.
+
+**Note:** when the static PVC created on the secondary site we cannot use the
+volumeHandle to identify the omap data or the image anymore because as we have
+only PoolID and clusterID in the volumeHandle. we cannot identify the correct
+pool name from the poolID because pool name will remain the same on both clusters
+but not the poolID even the clusterID can be different on the secondary
+cluster.
+
+> **VolumeHandle** is the unique volume name returned by the CSI volume plugin’s
+CreateVolume to refer to the volume on all subsequent calls.
+
+Once the static PVC is created on the secondary cluster, the user can try
+delete the PVC,expand the PVC or mount the PVC. Incase of mounting (node stage)
+we will get the volume context in GRPC call but not in the Delete/Expand
+Request. In Delete/Expand GRPC request only the
+volumeHandle (`clusterID-poolID-volumeuniqueID`) will be sent where it contains
+the encoded information related to clusterID and poolID. The volumeHandle is
+not useful in secondary cluster as the poolID and clusterID always may not be
+the same.
+
+To solve this problem, We will have a new controller running as part of
+provisioner pod which watchers for the PV objects. When a PV is created it will
+extract the required information from the PV spec and it will regenerate the
```

Comment:
> Few questions here:
> 
> *) Does this new controller run on both ( primary and secondary) clusters?
> 
yes it can run on both the clusters, functionality of the primary cluster wont be affected as the volumeHandler is already generated by CSI and omap data is already generated.

> *) Does this new controller run regardless of rbd mirroring is enabled in a setup ? iow, how are we controlling the deployment of this new controller from CSI deployment?
> 

It's up to the user to enable it or not, currently, it will be used only for mirroring use case but later the  user can make use of it later for expanding snapshot, and other operation which is not going to be covered here.


> *) What is the communication path between csi-rbdplugin container and the controller ( my assumption here is controller is just another container in the provisioner pod )? or it is going to be that, there is no communication between these, rather both works independently? if its independent, do we need to bundle it with provisioner pod?
> 
its nothing to do with the plugin as of now, it will run as a sidecar container in the provisioner pod.

> *) Is this new controller required, if CSI pod already know the `clusterID and Pool ID ` of the cluster ?
> 
It won't know anything, it will extract the information from the volumecontext

> *) It say `it will extract the required information` -> can we document what information is classified as `required` here?
> 
Sure I can update it not sure how much it will help, it will extract all the information present in the pv.spec.csi

> *) Controller expected to create a new VolumeHandle in format `newclusterID-newPoolID-volumeuniqueID` :
> 
> * Is it going to create a new volumehandle for all the `static PVs` in a cluster? ( regardless of whether the PV belong to mirrored image ?)
new volumehandle will not be used by the kubernetes, it will only be used by cephcsi internally for all operations.

> * Do we make use of certain properties of the PV to create this volumeUniqueID ? can you please list what are those properties if it tied ? or Is it  completely random UUID?
> 

It's the volumeHandle generated in the primary cluster.

> *) `whenever cephcsi gets a GRPC request with older volumeHandle it will check..` -> can you please confirm the check mentioned here is about the rbdplugin container?

yes rbdplugin will check the mapping between the volume handlers